File: examples/example-java-function/src/main/java/example/ListBucketsFunction.java
package example;

import com.oracle.bmc.objectstorage.ObjectStorageClient;
import com.oracle.bmc.objectstorage.model.BucketSummary;
import com.oracle.bmc.objectstorage.requests.GetNamespaceRequest;
import com.oracle.bmc.objectstorage.requests.ListBucketsRequest;
import io.micronaut.core.annotation.ReflectiveAccess;
import io.micronaut.oci.function.OciFunction;

import javax.inject.Inject;
import javax.inject.Singleton;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

@Singleton
public class ListBucketsFunction extends OciFunction {
    @Inject
    ObjectStorageClient objectStorageClient;

    @Inject
    TenantIdProvider tenantIdProvider;

    @ReflectiveAccess
    public List<String> handleRequest() {
        try {
            String tenancyId = tenantIdProvider.getTenantId();

            GetNamespaceRequest getNamespaceRequest = GetNamespaceRequest.builder()
                .compartmentId(tenancyId).build();
            String namespace = objectStorageClient.getNamespace(getNamespaceRequest).getValue();
            final ListBucketsRequest.Builder builder = ListBucketsRequest.builder();
            builder.namespaceName(namespace);
            builder.compartmentId(tenancyId);
            return objectStorageClient.listBuckets(builder.build())
                    .getItems().stream().map(BucketSummary::getName)
                    .collect(Collectors.toList());
        } catch (Throwable e) {
            System.out.println("Error occurred: " + e.getMessage());
            e.printStackTrace();
            return Collections.emptyList();
        }

    }
}

File: docs-examples/example-function-python/src/main/python/example/ListBucketsFunction.py
# tag::imports[]
from typing import Annotated

from com.oracle.bmc.objectstorage import ObjectStorageClient
from com.oracle.bmc.objectstorage.requests import GetNamespaceRequest, ListBucketsRequest
from jakarta.inject import Inject, Singleton
from micronaut.core.annotation import ReflectiveAccess
from micronaut.oraclecloud.core import TenancyIdProvider
from micronaut.oraclecloud.function import OciFunction
# end::imports[]


# tag::class[]
@Singleton
class ListBucketsFunction(OciFunction):  # <1>

    object_storage_client: Annotated[ObjectStorageClient, Inject]  # <2>

    tenant_id_provider: Annotated[TenancyIdProvider, Inject]
# end::class[]

    # tag::method[]
    @ReflectiveAccess
    def handle_request(self) -> list[str]:

        get_namespace_request = (GetNamespaceRequest.builder()
                                 .compartmentId(self.tenant_id_provider.getTenancyId()).build())
        namespace = self.object_storage_client.getNamespace(get_namespace_request).getValue()

        list_buckets_request = (ListBucketsRequest.builder()
                                .namespaceName(namespace)
                                .compartmentId(self.tenant_id_provider.getTenancyId())
                                .build())

        return [bucket.getName() for bucket in
                self.object_storage_client.listBuckets(list_buckets_request).getItems()]
    # end::method[]
